teps/0041-move-image-entrypoint-lookup-to-the-taskrun-pod.md

```diff
@@ -0,0 +1,161 @@
+
+Even though we are already in the image we want to query the entrypoint from
+after the architecture changing, the image metadata still needs to be fetched
+from `containerregistry`. The ideal solution should be fetching image config
+from the `kubelet` or underlying runtime straightly. However, the `kubelet` API
+that is capable to do this does not exist for now.
+
+### Add Flags to EntryPoint Binary
+
+The entrypoint binary should resolve the image entrypoint or command when the
+step container didn't specify a command. In order to find the image, three
+paramters could be passed to the binary - `image`, `taskrun_namespace`, and
+`taskrun_service_account`. `image` is used to reference the image.
+`taskrun_namespace` and `taskrun_service_account` are used to retrieve the
+authentication to pull the image.
+
+```go
+image = flag.String("image", "", "If specified, pull image and query image entrypoint")
+taskrunNamespace = flag.String("taskrun_namespace", "", "If image specified, use this namespace to pull image")
+taskrunServiceAccount = flag.String("taskrun_service_account", "", "If image specified, use this service account to pull image")
+```
+
+The parameter `entrypoint` must be empty when `image` is specified. The command
+to be run will look like this when step command isn't specified:
+
+```script
+/tekton/tools/entrypoint -wait_file '/tekton/downward/ready' -wait_file_content\
+-post_file '/tekton/tools/0' -termination_path '/tekton/termination'\
+-image '<image_path>' -taskrun_namespace 'default' -taskrun_service_account 'default'
+```
+
+### Resolve Image in EntryPoint Binary
+
+We will resolve entrypoint in the binary if `image` is specified. The steps list
+below:
+
+1. set up `kubeClient`
+1. get the keychains using `kubeClient`
+1. parse the image reference by the image name and pull the image from remote
+   `containerregistry` using the keychains
+1. extract the entrypoint from the image config, if entrypoint is not
+   specified, use the command instead
+1. add the entrypoint to `Entrypoint` struct and run the entrypoint
+   binary as usual
+
+### Notes/Caveats
+
+#### Entrypoint Cache
+
+(<em>If you have better ideas about cache, please feel free to complement</em>)
+
+Tekton controller maintains an `entrypointCache` when resolving the entrypoint.
+If the image can be found in the cache, there is no need to resolve it from the
+registry again. Besides `entrypointCache`, it also maintains `localCache` for
+single TaskRun, which means if multiple steps in one single TaskRun use the same
+image, it can be fetched from the `localCache` directly. This mechanism can
+reduce the startup latency time effectively.
+
+But it could be hard to maintain a image cache after moving the entrypoint
+lookup logic, because we resolve the entrypoint in different images. Some
+techniques like inter-pod communication or permanently storage must be involved
+to do this. Anyway, with or without cache, the cost of time will definitely be
```

Comment:
Could you go into more detail on how the image resolution cache would function when running on the pod (particularly how data would be shared between entrypoint invocations?)

Reply:
To implement cache for steps in the same pod, I plan to serialize the image and store it in the shared volume. One image will be serialized into one file. The filename is the image reference (whether the tag or the digest). When looking up the image entrypoint, we use the image reference to check if the file exists. if it does, grab it from the shared volume. Otherwise, fetch it from the remote registry.

Reply:
For the cache beyond the same pod, I could only think up a way that we need to bubble the whole image back to the controller and request the controller in every steps' container. But this way involves network communication between different pods. I'm not sure whether things would get faster if we introduced this. There is a big chance that consulting the remote registry directly costs less time compared to this kind of cache.

Reply:
Any way to reuse the `cache` part of `knative` (aka have an implementation of it) — I am fuzzy on the subject but felt it was worth mentionning.